saml/src/main/java/com/linecorp/armeria/server/saml/SamlServiceProviderBuilder.java

```diff
@@ -0,0 +1,479 @@
+
+    /**
+     * Sets a {@link SamlRequestIdManager} which creates and validates a SAML request ID.
+     */
+    public SamlServiceProviderBuilder requestIdManager(SamlRequestIdManager requestIdManager) {
+        this.requestIdManager = requireNonNull(requestIdManager, "requestIdManager");
+        return this;
+    }
+
+    /**
+     * Sets a {@link SamlSingleSignOnHandler} which handles SAML messages for a single sign-on.
+     */
+    public SamlServiceProviderBuilder ssoHandler(SamlSingleSignOnHandler ssoHandler) {
+        this.ssoHandler = requireNonNull(ssoHandler, "ssoHandler");
+        return this;
+    }
+
+    /**
+     * Sets a {@link SamlSingleLogoutHandler} which handles SAML messages for a single sign-on.
+     */
+    public SamlServiceProviderBuilder sloHandler(SamlSingleLogoutHandler sloHandler) {
+        this.sloHandler = requireNonNull(sloHandler, "sloHandler");
+        return this;
+    }
+
+    /**
+     * Returns a {@link SamlIdentityProviderConfigBuilder} to configure a new idp for authentication.
+     */
+    public SamlIdentityProviderConfigBuilder idp() {
+        final SamlIdentityProviderConfigBuilder config = new SamlIdentityProviderConfigBuilder(this);
+        idpConfigBuilders.add(config);
+        return config;
+    }
+
+    /**
+     * Returns a {@link SamlAssertionConsumerConfigBuilder} to configure a new assertion consumer service
+     * of this service provider.
+     */
+    public SamlAssertionConsumerConfigBuilder acs() {
+        final SamlAssertionConsumerConfigBuilder config = new SamlAssertionConsumerConfigBuilder(this);
+        acsConfigBuilders.add(config);
+        return config;
+    }
+
+    /**
+     * Builds a {@link SamlServiceProvider} which helps a {@link Server} have a SAML-based
+     * authentication.
+     */
+    public SamlServiceProvider build() {
+
+        // Must ensure that OpenSAML is initialized before building a SAML service provider.
+        SamlInitializer.ensureAvailability();
+
+        if (entityId == null) {
+            throw new IllegalArgumentException("entity ID is not specified");
+        }
+        if (credentialResolver == null) {
+            throw new IllegalArgumentException(CredentialResolver.class.getSimpleName() + " is not specified");
+        }
+        if (authorizer == null) {
+            throw new IllegalArgumentException(Authorizer.class.getSimpleName() + " is not specified");
+        }
+
+        final Credential signingCredential = credentialResolver.apply(signingKey);
+        if (signingCredential == null) {
+            throw new IllegalArgumentException("cannot resolve a " + Credential.class.getSimpleName() +
+                                               " for signing: " + signingKey);
+        }
+        final Credential encryptionCredential = credentialResolver.apply(encryptionKey);
+        if (encryptionCredential == null) {
+            throw new IllegalArgumentException("cannot resolve a " + Credential.class.getSimpleName() +
```

Comment:
Use `IllegalStateException` instead of `IllegalArgumentException` in `build()`.

Reply:
Fixed